### PR TITLE
Tsmith35 per BS PR4631

### PR DESCRIPTION
**Pull request by tsmith35**
# Description

Porting BS PR 4631 to OrcaSlicer.
[Bambu Studio PR#4631](https://github.com/bambulab/BambuStudio/pull/4631) was authored by @ziehmon

# Screenshots/Recordings/Graphs

Bambu Studio after PR4631 top, OrcaSlicer after ported changes bottom

BS-spiral

OS-spiral

### CODE REVIEW

**Comment by SoftFever** [comment]
Thank you.

Cloud you add the original author as the co author?

**Comment by tsmith35** [comment]
> Thank you.
> 
> Cloud you add the original author as the co author?

I made an additional commit (minor text changes) and added @ziehmon as co-author. Hopefully that will do it, but I added credit in the PR description as well.

**Comment by SoftFever** [comment]
Seems like Orca don't have this bug from the first place.
Following is the screenshot from Orca nightly build, seems all fine to me.
Can you verify?
image

**Comment by ziehmon** [comment]
@SoftFever: do reproduce this, you need to remove `timelapse_gcode` in extruder settings. This is because the nature of the bug: Bambu inserted the spiral lift inside the `timelapse_gcode`, but it won't be executed if timelapse is disabled. By removing the `timelapse_gcode`, you'll simulate gcode generation without timelapse enabled, since Bambu Lab/Orca assume it is enabled for the gcode preview. I reproduced this with nightly. 

Also by looking at the code, you can spot that Orca has this issue:

`gcode += this->retract(false, false, LiftType::NormalLift);` e.g. in https://github.com/SoftFever/OrcaSlicer/pull/6709/files#diff-cb3d09d2073b6e9032c65d07aa8f10d166324a53136b2aa1288f7b76f7d94319L4033, which is ultimately caused by: `m_writer.set_current_position_clear(false);` e.g. in https://github.com/SoftFever/OrcaSlicer/pull/6709/files#diff-cb3d09d2073b6e9032c65d07aa8f10d166324a53136b2aa1288f7b76f7d94319L4038

File: src/libslic3r/GCode.cpp
Comment (ziehmon) [inline src/libslic3r/GCode.cpp:6670]: It's just a comment, but I am curious what caused this diff?

File: src/libslic3r/GCode.cpp
Comment (ziehmon) [inline src/libslic3r/GCode.cpp:4156]: Did you changed the formatting of the gcode debug information for the object sequence on purpose? I imagine it could be a bit cleaner, just wondering. Was not part of the BambuStudio PR, but why not.

**Review by ziehmon — COMMENTED** [review]
I only spotted the two sections I commented to be different to the original PR, besides that, LGTM.

File: src/libslic3r/GCode.cpp
Comment (tsmith35) [reply src/libslic3r/GCode.cpp:6670]: I had seen the last comment formatted similarly in Bambu Studio's GCodeProcessor.hpp file, so I just carried it over to the files while I was there.

File: src/libslic3r/GCode.cpp
Comment (tsmith35) [reply src/libslic3r/GCode.cpp:4156]: BS inserts the debug comments in reverse order compared to OS. BS uses temp_start_str as a condition for adding **end_str += "M625\n;"**, which was my intent. I had some things come up that prevented me from getting things finished. I can probably just strike those parts out, though the original functionality remains as it was before. Never enough time.

Comment (ziehmon) [reply src/libslic3r/GCode.cpp:4156]: Thanks @tsmith35, that makes sense. No worries, you focused on the important changes and they look fine 👍🏻

**Comment by SoftFever** [comment]
@ziehmon Thank you for the explanation and code review!
Yeah, I can repro it after deleting the time-lapse custom gcode.

@tsmith35 All good for me. Thank you very much!

**Review by SoftFever — APPROVED** [review]

**Comment by ziehmon** [comment]
My pleasure @SoftFever, super nice to see the fix in Orca. Will update the reddit post about this and ask for some before/after photos. Also thanks again to @tsmith35 for porting!

**Comment by ziehmon** [comment]
Hi @SoftFever, hi @tsmith35,

unfortunately we need to revert this MR. There is a bug when AMS is used, since the position after changing filaments will be at X asis boundary and a spiral Z hop will move the axis out of physical boundaries. Details here: https://github.com/bambulab/BambuStudio/issues/4841#issuecomment-2362988198

I am working on fixing this and will provide the fix here too. Sorry, I should've bought an AMS...

**Comment by ziehmon** [comment]
New PR: https://github.com/bambulab/BambuStudio/pull/4848

Let's wait for review and testing. Will surely take some time, but at least that's fixed again...for now

**Comment by SoftFever** [comment]
> Hi @SoftFever, hi @tsmith35,
> 
> unfortunately we need to revert this MR. There is a bug when AMS is used, since the position after changing filaments will be at X asis boundary and a spiral Z hop will move the axis out of physical boundaries. Details here: [bambulab/BambuStudio#4841 (comment)](https://github.com/bambulab/BambuStudio/issues/4841#issuecomment-2362988198)
> 
> I am working on fixing this and will provide the fix here too. Sorry, I should've bought an AMS...


No sorry at all! 
You are doing angle's work here. 👍

**Comment by ziehmon** [comment]
> > Hi @SoftFever, hi @tsmith35,
> > unfortunately we need to revert this MR. There is a bug when AMS is used, since the position after changing filaments will be at X asis boundary and a spiral Z hop will move the axis out of physical boundaries. Details here: [bambulab/BambuStudio#4841 (comment)](https://github.com/bambulab/BambuStudio/issues/4841#issuecomment-2362988198)
> > I am working on fixing this and will provide the fix here too. Sorry, I should've bought an AMS...
> 
> No sorry at all! You are doing angle's work here. 👍

Hi @SoftFever,

I'm verry sorry to necro bump this, but I need your opinion and I think you are be the best person to ask right now. 

The reason why we had to rollback the change was that the injected filament change gcode from BambuStudio ends in an inconsistent and unsafe position at the bed boundary, instead of `x=0,y=0`. Other gcode (e.g. timelapse) ends at `x=0,y=0`.

I raised a PR to fix this behavior here https://github.com/bambulab/BambuStudio/pull/4848, which includes a full write up (it literally just moves the print head to the safe position `x=0,y=0` in filament change gcode). 

However, Bambu seems to be unwilling to merge it and asks for additional position validations. Weirdly, they ask to only validate non-injected gcode (so, invalid position data when injections happen, e.g. by filament change) and I can't make a sense out of it. Maybe it's communication issues, but anyway, not your project, not your beer.

Now, my question to you as a maintainer of Orca: would it be enough from your perspective to simply change the filament gcode to end in `x=0,y=0` and re-introduce the PR that adds the spiral Z hop? I am just asking for your thoughts, I'll implement this myself - promise.

If we can't fix this nasty stringing bug in BambuStudio, maybe we can do it in Orca.

**Pull request merged**
